AD4696x Spi Engine and Functional updates

pmallick-adi opened the pull request:
* Added macro so that the drivers can be used with normal MCU without spi engine.
* Extended the existing drivers functionality.

amiclaus reviewed (commented):
do we want to use this driver without spi engine?

Also, there are some build issue with the ad469x on your PR. Can you check please the logs and fix them?

pmallick-adi commented:
> do we want to use this driver without spi engine?
> 
> Also, there are some build issue with the ad469x on your PR. Can you check please the logs and fix them?

Yes, a lot of projects request that we get are using a MCU to interface with the AD4696 and not FPGA. 

I looked into the issue and made some changes. Should fix it hopefully

adrimbarean reviewed (commented):
Solid changes for the driver. A few questions an comments from me.

pmallick-adi commented on drivers/adc/ad469x/ad469x.c:
```diff
@@ -769,6 +802,73 @@ int32_t ad469x_read_data(struct ad469x_dev *dev,
+		return ret;
+
+	if(!((reset_status & AD469x_REG_STATUS_RESET_MASK) >> 5))
+		return ret;
```
Will fix this. Thanks.

pmallick-adi commented on drivers/adc/ad469x/ad469x.c:
```diff
@@ -803,11 +903,11 @@ int32_t ad469x_init(struct ad469x_dev **device,
 #endif
 
-	ret = ad469x_init_gpio(dev, init_param);
+	ret = no_os_spi_init(&dev->spi_desc, init_param->spi_init);
```
Actually, before creating a  ad469x_reset_dev() function, it was contained inside the ad469x_init_gpio() which would call the spi_read function read the status registers after reset. Since spi_init happens after it, so it shall result into a hard fault.
 
But now it is different function and I have updated the code accordingly.

pmallick-adi commented on drivers/adc/ad469x/ad469x.c:
```diff
@@ -229,23 +229,32 @@ static int32_t ad469x_init_gpio(struct ad469x_dev *dev,
 {
 	int32_t ret;
 
+	/** Initialize the RESET GPIO descriptor **/
```
Will remove these

pmallick-adi commented on drivers/adc/ad469x/ad469x.c:
```diff
@@ -229,23 +229,32 @@ static int32_t ad469x_init_gpio(struct ad469x_dev *dev,
-		ret = no_os_gpio_direction_output(dev->gpio_resetn, NO_OS_GPIO_LOW);
-		if (ret != 0)
-			return ret;
+	ret = no_os_gpio_direction_output(dev->gpio_resetn, NO_OS_GPIO_HIGH);
```
Yes we do need to toggle the pin low then high for the reset to happen. 
However that is being done now is the  ad469x_reset_dev function.

pmallick-adi commented on drivers/adc/ad469x/ad469x.c:
```diff
@@ -736,6 +757,14 @@ int32_t ad469x_read_data(struct ad469x_dev *dev,
 	if (dev->dcache_invalidate_range)
 		dev->dcache_invalidate_range(msg.rx_addr, samples * 4);
+#else
+	uint8_t reg_data[2] = {0};
+	ret = no_os_spi_write_and_read(dev->spi_desc, reg_data, (samples * 2));
+	if (ret != 0)
+		return ret;
+
+	*buf = (uint16_t)(reg_data[0] << 8) | reg_data[1];
```
I think this code shall not interfere with the spi_engine code because it come under #else block. The samples size from the ADC is 16-bit long.

adrimbarean replied:
What I'm trying to say is, what if `samples` is 100? Will it always be 2 if the SPI Engine is not used?

pmallick-adi replied:
I get your point. Will update this.

pmallick-adi replied:
Do you agree on this?

pmallick-adi commented:
Updated Log:

* Restored order of gpio and spi init function in the ad469x_init() function.
* Fixed the return type in ad469x_reset_dev() function.
* Remove the comments from the ad469x_init_gpio() function.

pmallick-adi commented:
Change Log:
* Uncommented the ENABLE_SPI_ENGINE macro for the project builds to pass.

adrimbarean commented on drivers/adc/ad469x/ad469x.c:
```diff
@@ -329,11 +338,28 @@ int32_t ad469x_std_seq_osr(struct ad469x_dev *dev, enum ad469x_osr_ratios ratio)
+	dev->std_seq_pin_pairing = pin_pair;
+
+	return ad469x_spi_reg_write(dev,
+				    AD469x_REG_CONFIG_IN(0),
```
Understood. Thank you!

pmallick-adi commented:
Any updates on this?

pmallick-adi commented:
Please reply on this. There is release dependent on these changes.

pmallick-adi commented:
Change Log:

Made of USE_STANDARD_SPI macro.

pmallick-adi commented:
Can we get this merged?

pmallick-adi commented:
Change Log:

Correct CYC_CTRL bit for setting into single cycle and two cycle mode referring to table 15 in the data sheet

adrimbarean reviewed (approved):
Looks good to me.
